### DISCUSSION TITLE
720

### DISCUSSION THREAD

**Discussion by giscus** [jojoldu/blog-comments#901] [Comments]
# 720

기존 SSR 시스템(jquery + Server Template)을 API 기반의 신규 시스템으로 전환하는 작업을 하면서 팀의 백엔드 개발자분들께 Response Body 관해서 종종 코멘트를 한다. 이런 코멘트들이 쌓이다보니 그냥 한번에 사내 위키로 정리하는게 좋겠단 생각에 먼저 블로그에 정리하게 되었다. 일반적인 내용들 + 팀 전용 내용들을 다 같이 담아서 사내 위키에 남기려고. 정답이 있는 것은 아니지만, 개인적으로 선호하는 Reponse Body 규칙들이 있다. API 역시 사용자 인터페이스이다. 인터페이스라는 측면을 무시하고, 내가 편한 방법, 혹은 기준 없이 API의 Body를 만들다보면 정작 사용해야하는 쪽에서 불편할때가 많다. (사용자 인터페이스임에도 불구하고 말이다) 내가 선호하는 API..

https://jojoldu.tistory.com/720

**Comment by donghL-dev** [comment]
좋은 글 잘 읽었습니다.

향로님 말씀처럼 정답이 없는 부분이긴 한데, 설명해주신 부분에서 질문이 있어서 댓글을 남깁니다.

`표준 포맷 사용하기` 부분에서 날짜 데이터 표기를 예를 들어서 설명해주셨는데, UNIX 타임 스탬프 형식으로 표기하면 프로그래머가 아니라면 알아보기 힘들다는 내용이 잘 이해되었습니다. 그런데 날짜 데이터 표기 시에 표준 시간대도 포함해서 표현해주면 호출해야 하는 쪽에서 데이터를 좀 더 직관적으로 이해할 수 있지 않을까 라는 생각이 들었습니다.

물론 서비스에 따라 이 날짜 데이터는 한국 서비스의 API 서버에서 내려주는 데이터니까, 한국 시간대 기준의 날짜 데이터로 이해하면 될거라고 생각할 수 있지만 실제로 한국 시간대라는 표기도 포함되어 있다면 좀 더 직관적으로 이해할 수 있지 않을까 라는 생각이 들었습니다.

API 응답 값의 날짜 데이터에 표준 시간대도 포함되면 그 날짜 데이터를 받는 쪽에서 보여주고자 하는 시간대로 변환하여 사용하면 좀 더 범용적으로 사용할 수 있지 않을까 라는 생각을 하였습니다. 

날짜 데이터 표기 시 표준 시간대 데이터를 포함하는 것에 대해서 질문 드리고자 댓글을 작성하게 되었습니다. 

(물론 `RFC 3339` 혹은 `ISO-8601`에 정의된 형식이라고 하신 부분에서 표준 시간대도 포함해서 말씀하신 거라면 제가 거기까지 인지하지 못하고 질문 드린 것일 수도 있습니다.)

 읽어주셔서 감사드립니다.

**Reply by kakaru1331** [reply]
HTTP API의 경우 Request header를 이용하는 게 좋을 거 같네요. 
API는 요청자 locale에 맞추어 날짜/시간을 반환하는 식으로요~

**Reply by JoMingyu** [reply]
ISO 8601 format에서는 UTC와의 time offset을 뒤에 표시하기 때문에 괜찮습니다.

예를 들면 `2023-01-23T13:55:08+0900` 은 UTC+9시간 기준 2023년 1월 23일 13시 55분 08초라는 의미이며, UTC로 환산하면 2023년 1월 23일 04시 55분 08초가 됩니다.

반면 본문에서는 뒤에 time offset이 빠져 있어서 혼란이 좀 있는 것도 같네요. UTC를 표현하려면 뒤에 `Z`를 붙이거나 하니까요. 무시하는 경우 표준 상으로는 UTC를 의미합니다.

**Reply by JoMingyu** [reply]
`Asia/Seoul` , `Asia/Tokyo` 같은 것을 사용한다면 인지 상 좋긴 하겠지만, timezone이라는 게 시간이 지남에 따라 변화무쌍해서... (time offset이 달라지는 경우도 있고, 서머타임이나, 새로운 timezone이 생기기도 하는 등)

라이브러리 업데이트가 뒤에 따라와야 할거라 오히려 혼란스러울 것 같습니다.

**Reply by donghL-dev** [reply]
답변 감사드립니다.

일단 제 댓글에 포함된 표준 시간대 데이터를 포함시키면 좋을 것 같다는 내용도 ISO 8601 포맷과 같이 time offset을 뒤에 표기하거나 UTC를 표현하는 `Z`를 붙이는 등의 표현이 추가되면 좋을 것 같다는 내용이긴 하였습니다. 말씀 주신것처럼 본문에서는 time offset이 빠져 있어서 혼란이 있었던 것 같아서 그 내용을 질문 드린 것이었습니다.

하지만 답글로 말씀 주신것처럼 아무런 표기가 되어 있지 않은 경우 표준 상으로는 UTC를 의미한다면 꼭 time offset을 표기함으로써 표준 시간대 데이터를 포함할 필요는 없을 것 같긴 하네요. 답변 감사드립니다.

**Comment by cremazer** [comment]
좋은내용 공유해주셔서 고맙습니다.

**Comment by sjquant** [comment]
저희 팀에서도 관련 내용 공유해보고 논의해봐야겠네요 :)
좋은 내용 감사드립니다!

**Comment by appkr** [comment]
Double형 데이터는 다들 어떻게 사용하시나요? 저는 JSON으로 (역)직렬화할땐 문자열 타입으로 설계합니다. 

(bad)
PUT /value {“key”: 33.0}
GET /value {“key”: 33.000000..1}

**Comment by jun1718** [comment]
좋은 글 잘 읽었습니다!
기존에 조금씩 헷갈리고 토론으로 이어지기도 했던 부분을 명쾌하게 정리해 주신 것 같습니다 :)

저는 신입 사원 역량을 가진 주니어 개발자입니다만,
추가로 두 가지 내용이 들어가면 어떨까 하는 의견이 궁금해서 댓글 남깁니다.

1) 숫자보다는 문자
response body로 bigint 혹은 자바의 Long 타입을 반환할 경우가 있을 수 있는데, 굉장히 커질 수 있는 이런 타입의 값은 문자열로 반환해 주는 것이 좋을 것 같습니다.
받는 쪽에서는 이를 Long 타입으로 받아야 할지 int 타입으로 받아야 할지 등의 고민도 필요 없을 것이구요! (대부분의 언어는 지원하겠지만 해당 숫자 범위 형식을 지원하지 않는 언어도 있을 수도 있다고 생각합니다)
이런 의미에서 보면 본문에서 코드값(enum)을 문자열 형태로 보내자고 하셨기 때문에 범위가 작은 타입의 숫자형도 필요 없어 보이고, 사실상 json에서 숫자형을 사용하는 것은 크게 의미가 없지 않을까 싶습니다. (간단히 책의 수량 등을 나타내는 경우만 숫자형으로 반환하면 되지 않을까 싶습니다)

2) 예상치 못한 PK
멤버의 pk값 같은 쉽게 노출되면 보안상 위험이 있어 보이는 pk들에 한해서는 auto_increment(mysql) 값을 사용하지 않고 내부 채번이나 여러 방식을 통해 예측 불가능한 값을 만들어주면 좋을것 같습니다. 대부분 알고 계시는 지식이지만 표준을 만들려고 하는 이런 문서에도 포함되면 좋지 않을까 싶습니다:)

**Comment by benelog** [comment]
camelCase로 API응답의 속성명을 쓰는 건 JSON(JavaScript Object Notation)이기 때문에 그래야하지 않을까 하는 생각을 평소에 하고 있었습니다.

JSON에 snake_case를 쓰는 사례는 Ruby On Rails와 이를 활용했던 GitHub.com, PHP 등의 영향으로 보이는데, 구현 언어는 나중에 바뀔 수 있으므로 JSON의 본래 컨벤션에 충실한 것이 미래를 위한 무난한 선택이지 않을까 싶습니다. Ruby나 PHP를 주로 쓰시는 분들의 생각도 궁금하네요.
